test/dataset.jl

```diff
@@ -0,0 +1,310 @@
+                )
+
+                # Test that we successfully extracted the flattened kwargs.
+                @test issetequal(
+                    [(:group1, :a), (:group1, :b), (:group2, :a), (:group2, :b)], keys(ds)
+                )
+
+                # Test an example where we don't use the default delimiter
+                ds = Dataset(; flatten(data, :_)...)
+
+                @test constraintmap(ds) == LittleDict(
+                    Pattern((:__, :time)) => Set([
+                        (:group1_a, :time),
+                        (:group1_b, :time),
+                        (:group2_a, :time),
+                        (:group2_b, :time),
+                    ]),
+                    Pattern((:__, :loc)) => Set([(:group1_a, :loc), (:group2_a, :loc)]),
+                    Pattern((:__, :obj)) => Set([(:group1_a, :obj), (:group2_a, :obj)]),
+                    Pattern((:__, :label)) => Set([(:group1_b, :label), (:group2_b, :label)]),
+                )
+
+                @test issetequal(
+                    [(:group1_a,), (:group1_b,), (:group2_a,), (:group2_b,)], keys(ds)
+                )
+            end
+            @testset "Pairs" begin
+                # In this case, we have the option to having tuple keys which allows for
+                # multi-dimensional style indexing.
+                data = [
+                    :group1 => [
+                        :a => KeyedArray(
+                            rand(4, 3, 2);
+                            time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                            loc=1:3,
+                            obj=[:a, :b],
+                        ),
+                        :b => KeyedArray(
+                            rand(4, 2);
+                            time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                            label=["x", "y"],
+                        )
+                    ],
+                    :group2 => [
+                        :a => KeyedArray(
+                            rand(4, 3, 2) .+ 1.0;
+                            time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                            loc=1:3,
+                            obj=[:a, :b],
+                        ),
+                        :b => KeyedArray(
+                            rand(4, 2) .+ 1.0;
+                            time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+                            label=["x", "y"],
+                        )
+                    ]
+                ]
+                ds = Dataset(flatten(data)...)
+
+                # Test that we successfully extracted the dims
+                @test issetequal([:time, :loc, :obj, :label], dimnames(ds))
+
+                # Test that we successfully extracted the flattened pairs as tuples.
+                @test issetequal(
+                    [(:group1, :a), (:group1, :b), (:group2, :a), (:group2, :b)], keys(ds)
```

Comment:
Ok, thanks that clears it up. I was confused by this comment

>In this case, we have the option to having tuple keys which allows for multi-dimensional style indexing" 

which lead me to  believe that this test was _only_ possible with the `Pairs` version.
But I see that the `NamedTuple` version would can still satisfy `keys(data) == [(:group1, :a), (:group1, :b), (:group2, :a), (:group2, :b)]` because it correctly splits them on the `⁻` delimiter.